consensus/src/dag/bootstrap.rs

```diff
@@ -1,136 +1,303 @@
+        let dag_handler = NetworkHandler::new(
+            self.epoch_state.clone(),
+            rb_handler,
+            dag_driver,
+            fetch_handler,
+            node_fetch_waiter,
+            certified_node_fetch_waiter,
+            state_sync_trigger,
+        );
+
+        (dag_handler, dag_fetcher)
+    }
+
+    async fn bootstrapper(
+        self,
+        mut dag_rpc_rx: Receiver<Author, IncomingDAGRequest>,
+        ordered_nodes_tx: UnboundedSender<OrderedBlocks>,
+        shutdown_rx: oneshot::Receiver<()>,
+    ) -> anyhow::Result<()> {
+        let adapter = Arc::new(NotificationAdapter::new(
+            ordered_nodes_tx,
+            self.storage.clone(),
+        ));
+
+        let sync_manager = DagStateSynchronizer::new(
+            self.epoch_state.clone(),
+            adapter.clone(),
+            self.time_service.clone(),
+            self.state_computer.clone(),
+            self.storage.clone(),
         );
-    let fetch_requester = Arc::new(fetch_requester);
 
-    let dag_driver = DagDriver::new(
+        // TODO: fetch the correct block info
+        let ledger_info = LedgerInfoWithSignatures::new(
+            LedgerInfo::new(BlockInfo::empty(), HashValue::zero()),
+            AggregateSignature::empty(),
+        );
+
+        let mut shutdown_rx = shutdown_rx.into_stream();
+
+        loop {
+            let (rebootstrap_notification_tx, mut rebootstrap_notification_rx) =
+                tokio::sync::mpsc::channel(1);
+
+            // TODO: fix
+            let (dag_store, order_rule) =
+                self.bootstrap_dag_store(0, ledger_info.ledger_info().clone(), adapter.clone());
+
+            let state_sync_trigger = StateSyncTrigger::new(
+                dag_store.clone(),
+                adapter.clone(),
+                rebootstrap_notification_tx,
+            );
+
+            let (handler, fetch_service) =
+                self.bootstrap_components(dag_store.clone(), order_rule, state_sync_trigger);
+
+            let df_handle = tokio::spawn(fetch_service.start());
+
+            // poll the network handler while waiting for rebootstrap notification or shutdown notification
+            select! {
+                biased;
+                _ = shutdown_rx.select_next_some() => {
+                    df_handle.abort();
+                    let _ = df_handle.await;
+                    return Ok(());
+                },
+                Some(node) = rebootstrap_notification_rx.recv() => {
+                    df_handle.abort();
+                    let _ = df_handle.await;
+
+                    let dag_fetcher = DagFetcher::new(self.epoch_state.clone(), self.dag_network_sender.clone(), self.time_service.clone());
+
+                    if let Err(e) = sync_manager.sync_dag_to(&node, dag_fetcher, dag_store.clone()).await {
+                        error!(error = ?e, "unable to sync");
+                    }
+                },
+                _ = handler.start(&mut dag_rpc_rx) => {}
```

Comment:
I don't think this works? once it starts, it'll never go out from the inner loop

Reply:
i think it works. as long as we have an `.await` within the loop, the future should stop polling and yield to another future in the outer select!. If we get rebootstrap notification, it will complete the future and then we will simply not poll handler again. instead, we will start a new handler.